### PR TITLE
*: rewrite builtin function: CURRENT_TIME

### PR DESCRIPTION
**Pull request by XuHuaiyu**
to #4080 
PTAL @zz-jason @breeswish

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1388,31 +1389,49 @@ func (c *currentTimeFunctionClass) getFunction(args []Expression, ctx context.Co
...
-	baseBuiltinFunc
+type builtinCurrentTime0ArgSig struct {
+	baseDurationBuiltinFunc
 }
 
-// eval evals a builtinCurrentTimeSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_curtime
-func (b *builtinCurrentTimeSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinCurrentTime0ArgSig) evalDuration(row []types.Datum) (types.Duration, bool, error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: move the comment back

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1388,31 +1389,49 @@ func (c *currentTimeFunctionClass) getFunction(args []Expression, ctx context.Co
...
 	}
-	sig := &builtinCurrentTimeSig{newBaseBuiltinFunc(args, ctx)}
+
+	if len(args) == 0 {
+		bf := newBaseBuiltinFuncWithTp(args, ctx, tpDuration)
+		bf.tp.Flen, bf.tp.Decimal, bf.deterministic = mysql.MaxDurationWidthNoFsp, types.MinFsp, false
+		sig := &builtinCurrentTime0ArgSig{baseDurationBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
+	// args[0] must be an in constant which will not be null.
```
Comment (zz-jason) [inline expression/builtin_time.go]: if you make assumptions that `arg[0]` is a `Constant`, you should do the type assertion before calling `args[0].EvalInt`

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1388,31 +1389,49 @@ func (c *currentTimeFunctionClass) getFunction(args []Expression, ctx context.Co
...
+type builtinCurrentTime0ArgSig struct {
+	baseDurationBuiltinFunc
 }
 
-// eval evals a builtinCurrentTimeSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_curtime
-func (b *builtinCurrentTimeSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinCurrentTime0ArgSig) evalDuration(row []types.Datum) (types.Duration, bool, error) {
+	res, err := types.ParseDuration(time.Now().Format("15:04:05.000000"), types.MinFsp)
```
Comment (zz-jason) [inline expression/builtin_time.go]: - can we change `15:04:05.000000` to `00:00:00.000000` ?
- if `err` is not nil, we should return `true`

Comment (XuHuaiyu) [reply expression/builtin_time.go]: 1. Noop, according to golang manual:
> Layouts must use the reference time Mon Jan 2 15:04:05 MST 2006 to show the pattern with which to format/parse a given time/string.

File: expression/builtin_time.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/builtin_time.go]: Here args[0] can only be an int constant for current_time which is promised be parser.

File: expression/builtin_time.go (hunk 3 above)
Comment (zz-jason) [reply expression/builtin_time.go]: then we should extract this magic string as a constant

File: expression/builtin_time.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_time.go]: I see. But we still need to check the `NULL` value here. Once parser changes its behavior and not guarantee the parameter to be a constant or maybe parser passes a NULL constant, we can detect this change and make the right move.

**Comment by jackysp** [comment]
LGTM

File: expression/builtin_time_test.go [hunk 4]
```diff
@@ -1032,13 +1034,9 @@ func (s *testEvaluatorSuite) TestCurrentTime(c *C) {
 	c.Assert(n.String(), GreaterEqual, last.Format(tfStr))
 
 	f, err = fc.getFunction(datumsToConstants(types.MakeDatums(-1)), s.ctx)
-	c.Assert(err, IsNil)
-	_, err = f.eval(nil)
 	c.Assert(err, NotNil)
```
Comment (zz-jason) [inline expression/builtin_time_test.go]: check the error message ?

**Comment by breeswish** [comment]
Please resolve conflicts

File: expression/bench_func_test.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
+package expression
```
Comment (zz-jason) [inline expression/bench_func_test.go]: mistaken included ?

**Comment by XuHuaiyu** [comment]
ptal @zz-jason @breeswish

File: expression/builtin_time.go [hunk 6]
```diff
@@ -1480,35 +1481,64 @@ type currentTimeFunctionClass struct {
...
 	}
-	sig := &builtinCurrentTimeSig{newBaseBuiltinFunc(args, ctx)}
+
+	if len(args) == 0 {
+		bf := newBaseBuiltinFuncWithTp(args, ctx, tpDuration)
+		bf.tp.Flen, bf.tp.Decimal, bf.foldable = mysql.MaxDurationWidthNoFsp, types.MinFsp, false
+		sig = &builtinCurrentTime0ArgSig{baseDurationBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
+	// args[0] must be an in constant which will not be null.
```
Comment (zz-jason) [inline expression/builtin_time.go]: s/args[0] must be an in constant which will not be null/args[0] must be a constant which should not be null/

**Comment by zz-jason** [comment]
@XuHuaiyu plz resolve conflicts

File: expression/builtin_time.go [hunk 7]
```diff
@@ -1480,35 +1481,64 @@ type currentTimeFunctionClass struct {
...
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_curtime
+func (c *currentTimeFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinCurrentTimeSig{newBaseBuiltinFunc(args, ctx)}
+
+	if len(args) == 0 {
+		bf := newBaseBuiltinFuncWithTp(args, ctx, tpDuration)
+		bf.tp.Flen, bf.tp.Decimal, bf.foldable = mysql.MaxDurationWidthNoFsp, types.MinFsp, false
```
Comment (hanfei1991) [inline expression/builtin_time.go]: current time should be folded

**Comment by shenli** [comment]
LGTM

**Pull request merged**